src/coreComponents/constitutive/fluid/multifluid/compositional/CompositionalMultiphaseFluidUpdates.hpp

```diff
@@ -0,0 +1,431 @@
+  m_phase1( phase1.createKernelWrapper()),
+  m_phase2( phase2.createKernelWrapper()),
+  m_phase3( phase3.createKernelWrapper())
+{}
+
+template< typename FLASH, typename PHASE1, typename PHASE2, typename PHASE3 >
+GEOS_HOST_DEVICE
+GEOS_FORCE_INLINE
+void
+CompositionalMultiphaseFluidUpdates< FLASH, PHASE1, PHASE2, PHASE3 >::compute(
+  real64 const pressure,
+  real64 const temperature,
+  arraySlice1d< real64 const, compflow::USD_COMP - 1 > const & composition,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseFrac,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseDens,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseMassDens,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseVisc,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseEnthalpy,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseInternalEnergy,
+  arraySlice2d< real64, multifluid::USD_PHASE_COMP - 2 > const & phaseCompFrac,
+  real64 & totalDens ) const
+{
+  GEOS_UNUSED_VAR( phaseEnthalpy, phaseInternalEnergy );
+  GEOS_UNUSED_VAR( pressure, temperature );
+
+  integer constexpr maxNumComp = MultiFluidBase::MAX_NUM_COMPONENTS;
+  integer constexpr maxNumPhase = MultiFluidBase::MAX_NUM_PHASES;
+  integer const numComp = numComponents();
+  integer const numPhase = numPhases();
+
+  // 1. Convert input mass fractions to mole fractions and keep derivatives
+
+  stackArray1d< real64, maxNumComp > compMoleFrac( numComp );
+  if( m_useMass )
+  {
+    convertToMoleFractions< maxNumComp >( composition, compMoleFrac );
+  }
+  else
+  {
+    for( integer ic = 0; ic < numComp; ++ic )
+    {
+      compMoleFrac[ic] = composition[ic];
+    }
+  }
+
+  // 2. Compute phase fractions and phase component fractions
+  m_flash.compute( pressure,
+                   temperature,
+                   compMoleFrac.toSliceConst(),
+                   phaseFrac,
+                   phaseCompFrac );
+
+  // 3. Calculate the phase densities
+  m_phase1.density.compute( pressure,
+                            temperature,
+                            phaseCompFrac[0].toSliceConst(),
+                            phaseDens[0],
+                            phaseMassDens[0],
+                            m_useMass );
+  m_phase2.density.compute( pressure,
+                            temperature,
+                            phaseCompFrac[1].toSliceConst(),
+                            phaseDens[1],
+                            phaseMassDens[1],
+                            m_useMass );
+  if constexpr (2 < FLASH::getNumberOfPhases())
+  {
+    m_phase3.density.compute( pressure,
+                              temperature,
+                              phaseCompFrac[2].toSliceConst(),
+                              phaseDens[2],
+                              phaseMassDens[2],
+                              m_useMass );
+  }
+
+  // 4. Calculate the phase viscosities
+  m_phase1.viscosity.compute( pressure,
```

Comment:
The LBC viscosity model will require the phaseDensity as an argument as well.  Can put this now or wait until the LBC PR.

Reply:
I will add this now.